### PR TITLE
HAI-1964 Add hanke attachment migrator

### PR DESCRIPTION
**Pull request by pitkni**
# Description

Migrate hanke attachments from the database to cloud. Handle one attachment at a time.

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [x] Other

# Instructions for testing
1. Make sure to have some data in hanke_attachment_content table. 
2. Set HAITATON_ATTACHMENT_MIGRATION_HANKE_ENABLED: true in docker-compose.yml
3. Start whole haitaton stack with docker compose (and of course use a build of this branch).
4. Verify process from the logs, make sure it is readable and there are no errors.
5. Check db hanke_attachment_table, blob location should be updated with a value.
6. Migrated attachments should have been removed hanke_attachment_content table.
7. Hanke attachments can be viewed from the UI.

# Checklist:

- [x] I have written new tests (if applicable)
- [x] I have ran the tests myself (if applicable)
- [ ] I have made necessary changes to the documentation, link to confluence
 or other location: 

# Other relevant info
Please describe here if there is e.g. some requirements for this change or
 other info that the tester/user needs to know.

### CODE REVIEW

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigratorITest.kt [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+class HankeAttachmentMigratorITest(
+    @Autowired private val migrator: HankeAttachmentMigrator,
+    @Autowired private val attachmentFactory: HankeAttachmentFactory,
+    @Autowired private val fileClient: MockFileClient
+) : DatabaseTest() {
+
+    @Nested
+    inner class MigrationOperation {
+        @Test
+        fun `Should migrate to cloud and remove old content`() {
```
Comment (corvidian) [inline services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigratorITest.kt:49]: This test seems to test the `migrate()`-function in the test file that emulates the operation that the scheduler performs. There are no guarantees that the emulation will match the actual implementation. Why not test the scheduler function directly (in a separate test file)?

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        attachment.let { (id, hankeId, content) ->
+            val blobPath = HankeAttachmentContentService.generateBlobPath(hankeId)
+            fileClient.upload(
+                container = HANKE_LIITTEET,
+                path = blobPath,
+                originalFilename = content.fileName,
+                contentType = AttachmentValidator.parseMediaType(content.contentType),
+                content = content.bytes,
+            )
+            return MigrationResult(attachmentId = id, blobPath = blobPath)
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt:56]: Returning the attachment ID seems unnecessary since the caller already knows it. It's just copied from the call parameter to the response. This seems to be the only place the MigrationResult is used, so that class can be removed.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/common/AttachmentContentEntity.kt [hunk 3]
```diff
@@ -36,7 +37,13 @@ class ApplicationAttachmentContentEntity(
 ) : AttachmentContentEntity(attachmentId, content)
 
 @Repository
-interface HankeAttachmentContentRepository : JpaRepository<HankeAttachmentContentEntity, UUID>
+interface HankeAttachmentContentRepository : JpaRepository<HankeAttachmentContentEntity, UUID> {
+
+    @Query(value = "SELECT * FROM hanke_attachment_content LIMIT 1", nativeQuery = true)
+    fun pickOne(): HankeAttachmentContentEntity?
+
+    fun deleteByAttachmentId(attachmentId: UUID)
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/common/AttachmentContentEntity.kt:45]: Since attachmentId is the primary key, this is the same as the built-in `deleteById(ID id)`.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/common/AttachmentContentEntity.kt [hunk 4]
```diff
@@ -36,7 +37,13 @@ class ApplicationAttachmentContentEntity(
 ) : AttachmentContentEntity(attachmentId, content)
 
 @Repository
-interface HankeAttachmentContentRepository : JpaRepository<HankeAttachmentContentEntity, UUID>
+interface HankeAttachmentContentRepository : JpaRepository<HankeAttachmentContentEntity, UUID> {
+
+    @Query(value = "SELECT * FROM hanke_attachment_content LIMIT 1", nativeQuery = true)
+    fun pickOne(): HankeAttachmentContentEntity?
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/common/AttachmentContentEntity.kt:43]: This is equivalent to the derived query `findFirstBy()`.
```suggestion
    fun findFirstBy(): HankeAttachmentContentEntity?
```

File: services/hanke-service/src/main/resources/application.properties [hunk 5]
```diff
@@ -19,59 +18,44 @@ haitaton.gdpr.audience=${HAITATON_GDPR_AUDIENCE:http://localhost:8080/haitaton}
...
 management.endpoint.health.probes.enabled=true
 management.health.livenessState.enabled=true
 management.health.readinessState.enabled=true
 # Spring Boot Actuator Info properties
-info.java-version = ${java.specification.version}
-info.java-vendor = ${java.specification.vendor}
+info.java-version=${java.specification.version}
+info.java-vendor=${java.specification.vendor}
 server.forward-headers-strategy=FRAMEWORK
-
```
Comment (corvidian) [inline services/hanke-service/src/main/resources/application.properties:56]: I don't think removing empty lines from between sections is helping the readability.

File: services/hanke-service/src/main/resources/application.properties [hunk 6]
```diff
@@ -1,15 +1,14 @@
 haitaton.cors.allowedOrigins=${HAITATON_CORS_ALLOWED_ORIGINS:http://localhost:3001}
-
+# Allu properties
```
Comment (corvidian) [inline services/hanke-service/src/main/resources/application.properties:2]: If we want to divide the properties into strict sections, maybe a YAML file would be better?

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+@Component
+class HankeAttachmentMigrator(
+    private val fileClient: FileClient,
+    private val attachmentRepository: HankeAttachmentRepository,
+    private val contentRepository: HankeAttachmentContentRepository
+) {
+    /** Find an attachment that is not yet migrated, i.e. still has content in the db. */
+    @Transactional(readOnly = true)
+    fun unMigratedAttachment(): UnMigratedHankeAttachment? {
+        logAttachmentAmountToMigrate()
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt:30]: This is not related to getting an unmigrated attachment. It could be called separately from the Scheduler.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+@Component
+class HankeAttachmentMigrator(
+    private val fileClient: FileClient,
+    private val attachmentRepository: HankeAttachmentRepository,
+    private val contentRepository: HankeAttachmentContentRepository
+) {
+    /** Find an attachment that is not yet migrated, i.e. still has content in the db. */
+    @Transactional(readOnly = true)
+    fun unMigratedAttachment(): UnMigratedHankeAttachment? {
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt:29]: Unmigrated is one word.

A name that describes what this does would be better. Like `findAttachmentWithDatabaseContent()`.

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigratorITest.kt (hunk 1 above)
Comment (pitkni) [reply services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigratorITest.kt]: Done.

File: services/hanke-service/src/main/resources/application.properties (hunk 5 above)
Comment (pitkni) [reply services/hanke-service/src/main/resources/application.properties]: Added line breaks back.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt (hunk 7 above)
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt:30]: That would mean another dependency for the scheduler. I see this as part of the migration process so it should be in this class.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt (hunk 8 above)
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt]: Renamed.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/common/AttachmentContentEntity.kt (hunk 3 above)
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/common/AttachmentContentEntity.kt]: Oh, ok I did not see that. Thx.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt (hunk 7 above)
Comment (corvidian) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt:29]: This can be done without adding a dependency to the scheduler. `logAttachmentAmountToMigrate()` could be made public and called from the scheduler. Both classes are migration components, so which class handles which part is a pragmatic issue.
```kotlin
    fun scheduleMigrate() =
        lockService.doIfUnlocked(MIGRATE_HANKE_ATTACHMENT) {
            hankeAttachmentMigrator.logAttachmentAmountToMigrate()
            hankeAttachmentMigrator.findAttachmentWithDatabaseContent()?.let { migrate(it) }
                ?: logger.info { "No hanke attachments to migrate" }
        }
```
Then again, this is not a critical issue, especially since this code is short-lived.

File: services/hanke-service/src/main/resources/application.properties (hunk 5 above)
Comment (corvidian) [reply services/hanke-service/src/main/resources/application.properties]: All the useful edits were removed as well.

**Review by corvidian — APPROVED** [review]

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt (hunk 7 above)
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt:29]: Oh, ok. I misunderstood that the unMigratedAttachment should be moved to the scheduler.

Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/hanke/HankeAttachmentMigrator.kt:29]: The suggestion would be more logical I guess, but I think it would mean two different transactions. As said, this is a short lived code, so I'll leave it as it was.
